File: final/pb17.py
import sys
from collections import deque

inc_x = [ 1,  1, -1, -1 , 2,  2, -2, -2 ]
inc_y = [ 2, -2,  2, -2 , 1, -1,  1, -1 ]

f = 0
c = 0
n = 0
T = []
G = []

def veins(x, y):
    global f, c, n, T, G
    if T[x][y]:
        return
    v = c*x + y
    for k in range(8):
        x2 = x + inc_x[k]
        y2 = y + inc_y[k]
        if (x2 >= 0 and x2 < f and y2 >= 0 and y2 < c and not T[x2][y2]):
            G[v].append(c*x2 + y2)

def amplada(v):
    global f, c, n, T, G
    dist = [-1] * n
    Q = deque()
    dist[v] = 0
    Q.append(v)
    res = -33
    while len(Q) > 0:
        u = Q.popleft()
        for z in G[u]:
            if dist[z] == -1:
                res = dist[u] + 1
                dist[z] = res
                Q.append(z)
    return res

def main():
    global f, c, n, T, G

    for line in sys.stdin:
        f, c = map(int, line.split())
        TT = [False] * c
        T  = [None]  * f
        for i in range(f):
            T[i] = TT.copy()

        entrada = list(map(int, input().split()))
        for i in range(1, 2 * entrada[0] + 1, 2):
            T[entrada[i] - 1][entrada[i+1] - 1] = True

        n = f*c

        G = [None] * n
        for i in range(n):
            G[i] = []

        for x in range(f):
            for y in range(c):
                veins(x,y)

        res = 0
        for v in range(n):
            res = max(res, amplada(v))

        print(res)

main()

File: final/pb17.cpp
#include <iostream>
#include <queue>
#include <vector>
using namespace std;


using VE = vector<int>;
using VVE = vector<VE>;


const VE inc_x = { 1,  1, -1, -1 , 2,  2, -2, -2 };
const VE inc_y = { 2, -2,  2, -2 , 1, -1,  1, -1 };


int f, c;
VVE T;
int n;
VVE G;


void veins(int x, int y) {
  if (T[x][y]) return;
  int v = c*x + y;
  for (int k = 0; k < 8; ++k) {
    int x2 = x + inc_x[k];
    int y2 = y + inc_y[k];
    if (x2 >= 0 and x2 < f and y2 >= 0 and y2 < c and not T[x2][y2])
      G[v].push_back(c*x2 + y2);
  }
}


int amplada(int v) {
  VE dist(n, -1);
  queue<int> Q;
  dist[v] = 0;
  Q.push(v);
  int res = -33;
  while (not Q.empty()) {
    int u = Q.front(); Q.pop();
    for (int z : G[u])
      if (dist[z] == -1) {
        res = dist[z] = dist[u] + 1;
        Q.push(z);
      }
  }
  return res;
}


int main() {
  int p;
  while (cin >> f >> c >> p) {
    T = VVE(f, VE(c, false));
    while (p--) {
      int x, y;
      cin >> x >> y;
      T[x-1][y-1] = true;
    }

    n = f*c;
    G = VVE(n);
    for (int x = 0; x < f; ++x)
      for (int y = 0; y < c; ++y) veins(x, y);

    int res = 0;
    for (int v = 0; v < n; ++v) res = max(res, amplada(v));
    cout << res << endl;
  }
}
